### ISSUE TITLE
docker-gc may lead to build errors

### ISSUE THREAD

**Issue by joshma** [buildkite/elastic-ci-stack-for-aws#321]
I currently suspect docker-gc, but don't have a reliable repro yet. However, `docker system prune -f` looks suspect (https://github.com/buildkite/elastic-ci-stack-for-aws/blob/master/packer/conf/docker/cron.daily/docker-gc#L5), and I think there may be a race condition.

The background here is I've been trying to troubleshoot why I see something like this in a build:
```
$ docker-compose -f docker-compose.buildkite.yml -p buildkite682c3781b7834be6ab0be49db0c04b73 run tests .buildkite/steps/test_integration.sh
Creating network "buildkite682c3781b7834be6ab0be49db0c04b73_default" with the default driver
Pulling redis (redis:latest)...
latest: Pulling from library/redis
...etc...
...etc...
...etc...
Removing intermediate container dc19d56fce9a
Step 11/11 : CMD elasticsearch
 ---> Running in 13647b9efced
 ---> 1283cf72ca16
Removing intermediate container 13647b9efced
Successfully built 1283cf72ca16
Successfully tagged benchling/sightglass:latest
WARNING: Image for service sightglass was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.
Creating buildkite682c3781b7834be6ab0be49db0c04b73_redis_1 ...
Creating buildkite682c3781b7834be6ab0be49db0c04b73_db_1 ...
Creating buildkite682c3781b7834be6ab0be49db0c04b73_rabbitmq_1 ...
Creating buildkite682c3781b7834be6ab0be49db0c04b73_sightglass_1 ...
Creating buildkite682c3781b7834be6ab0be49db0c04b73_redis_1
Creating buildkite682c3781b7834be6ab0be49db0c04b73_db_1
Creating buildkite682c3781b7834be6ab0be49db0c04b73_warehouse_1 ...
Creating buildkite682c3781b7834be6ab0be49db0c04b73_sightglass_1
Creating buildkite682c3781b7834be6ab0be49db0c04b73_rabbitmq_1
Creating buildkite682c3781b7834be6ab0be49db0c04b73_db_1 ... error
 
Creating buildkite682c3781b7834be6ab0be49db0c04b73_redis_1 ... error
 
Creating buildkite682c3781b7834be6ab0be49db0c04b73_warehouse_1 ... error
 
Creating buildkite682c3781b7834be6ab0be49db0c04b73_sightglass_1 ... error
 
Creating buildkite682c3781b7834be6ab0be49db0c04b73_rabbitmq_1 ... error
 
ERROR: for buildkite682c3781b7834be6ab0be49db0c04b73_rabbitmq_1  Cannot start service rabbitmq: network buildkite682c3781b7834be6ab0be49db0c04b73_default not found
ERROR: for warehouse  Cannot start service warehouse: network buildkite682c3781b7834be6ab0be49db0c04b73_default not found
 
ERROR: for redis  Cannot start service redis: network buildkite682c3781b7834be6ab0be49db0c04b73_default not found
 
ERROR: for db  Cannot start service db: network buildkite682c3781b7834be6ab0be49db0c04b73_default not found
 
ERROR: for rabbitmq  Cannot start service rabbitmq: network buildkite682c3781b7834be6ab0be49db0c04b73_default not found
 
ERROR: for sightglass  Cannot start service sightglass: network buildkite682c3781b7834be6ab0be49db0c04b73_default not found
```

Note that this is similar to #148, except there's no error around creating a network - it seems to succeed, until it tries to create containers and then it errors.

I was able to repro this behavior by running `docker system prune -f` in a shell right after I started running `docker-compose up --build` in a second shell. It looks like `docker system prune -f` doesn't think that network is in use yet, so it removes it, even though `docker-compose` wants to create containers on it later.

The changes that `docker system prune -f` was running exactly during that point in time are pretty low, but if a bunch of builds are happening it may be possible. (Maybe there's something here that correlates it? Maybe a new instance runs its `cron.d/` tasks asap, just as a build starts on it?) Either way, though, it seems dangerous to do.

Wanted to file this issue to get it on the radar - can we consider replacing the current `docker-gc` script with something less aggressive? I'm not sure what state of the art here is, for example maybe https://github.com/spotify/docker-gc could work?

**Comment by lox** [comment]
Interesting. It's entirely possible this is happening. It only runs daily, so it's impressive that you are running into this race condition. 

It seems like the safe thing to do would be to gracefully shut down the agent and then do the prune. If there is only 1 agent per-host then running the prune in an environment hook like in https://github.com/buildkite/elastic-ci-stack-for-aws/pull/317 should work too.

**Comment by lox** [comment]
We absolutely could consider going back to `docker-gc` too.

**Comment by lox** [comment]
Should be fixed in #326, will make it into the next release candidate.
